src/query/v2/multiframe.hpp

```diff
@@ -0,0 +1,276 @@
+  state in the multiframe.
+  Iteration goes in a deterministic order.
+  One can't modify the validity of the frame nor its content with this implementation.
+  */
+  ValidFramesReader GetValidFramesReader();
+
+  /*!
+  Returns a object on which one can iterate in a for-loop. By doing so, you will only get frames that are in a valid
+  state in the multiframe.
+  Iteration goes in a deterministic order.
+  One can't modify the validity of the frame with this implementation. One can modify its content.
+  */
+  ValidFramesModifier GetValidFramesModifier();
+
+  /*!
+  Returns a object on which one can iterate in a for-loop. By doing so, you will only get frames that are in a valid
+  state in the multiframe.
+  Iteration goes in a deterministic order.
+  One can modify the validity of the frame with this implementation.
+  If you do not plan to modify the validity of the frames, use GetValidFramesReader/GetValidFramesModifer instead as
+  this is faster.
+  */
+  ValidFramesConsumer GetValidFramesConsumer();
+
+  /*!
+  Returns a object on which one can iterate in a for-loop. By doing so, you will only get frames that are in an invalid
+  state in the multiframe.
+  Iteration goes in a deterministic order.
+  One can modify the validity of the frame with this implementation.
+  */
+  InvalidFramesPopulator GetInvalidFramesPopulator();
+
+  void ResetAllFramesInvalid() noexcept;
+
+  bool HasValidFrame() const noexcept;
+
+  inline utils::MemoryResource *GetMemoryResource() { return frames_[0].GetMemoryResource(); }
+
+ private:
+  void DefragmentValidFrames();
+
+  FrameWithValidity default_frame_;
+  utils::pmr::vector<FrameWithValidity> frames_ =
+      utils::pmr::vector<FrameWithValidity>(0, FrameWithValidity{1}, utils::NewDeleteResource());
+};
+
+class ValidFramesReader {
+ public:
+  explicit ValidFramesReader(MultiFrame &multiframe);
+
+  ~ValidFramesReader() = default;
+  ValidFramesReader(const ValidFramesReader &other) = delete;                 // copy constructor
+  ValidFramesReader(ValidFramesReader &&other) noexcept = delete;             // move constructor
+  ValidFramesReader &operator=(const ValidFramesReader &other) = delete;      // copy assignment
+  ValidFramesReader &operator=(ValidFramesReader &&other) noexcept = delete;  // move assignment
+
+  struct Iterator {
+    using iterator_category = std::forward_iterator_tag;
+    using difference_type = std::ptrdiff_t;
+    using value_type = const Frame;
+    using pointer = value_type *;
+    using reference = const Frame &;
+    using internal_ptr = FrameWithValidity *;
+
+    Iterator(internal_ptr ptr, ValidFramesReader &iterator_wrapper) : ptr_(ptr), iterator_wrapper_(iterator_wrapper) {}
+
+    reference operator*() const { return *ptr_; }
+    pointer operator->() { return ptr_; }
+
+    // Prefix increment
+    Iterator &operator++() {
+      do {
+        ptr_++;
+      } while (!this->ptr_->IsValid() && *this != iterator_wrapper_.end());
```

Comment:
> The order of the conditions should be reversed. This way it is possible that we dereference and end iterator which can point after the last element, so it might result in an invalid memory access.

This is true, I have updated all of them (regardless of the second remark)

Reply:
Heads up: I've implemented a version where we only need to know about the "after last valid frame"
Given that valid frames are at the beginning of the vector, we don't need to keep the first valid frame